### PR TITLE
Submitting assignment 1 for Backend

### PR DESCRIPTION
**Pull request by siddhesh2309**
IPL Model based on JAVA and Object Oriented Programming Concepts

### CODE REVIEW

File: assignment-1/Functions.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toMap;
+
+public class Functions {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:7]: Should have used more meaningful class name, This does not give any information about the functionality of the class.

File: assignment-1/Functions.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+    public static final String delimiter = ",";
+    public static ArrayList<Players> allPlayers = new ArrayList<Players>();
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:10]: Why is it declared as static?

File: assignment-1/Team.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import java.util.ArrayList;
+
+public class Team extends Players {
```
Comment (ameyaditya0408) [inline assignment-1/Team.java:3]: 1. Why does team extend player? 
2. did not see implementation  of hashcode and equals method.

File: assignment-1/Team.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    public String Name;
```
Comment (ameyaditya0408) [inline assignment-1/Team.java:4]: 1. never declare instance variables as public, always make their private and use getter methods
2. follow the proper naming convention, variable, and method names must begin with a lowercase character.

File: assignment-1/Team.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    public ArrayList<Players> players;
```
Comment (ameyaditya0408) [inline assignment-1/Team.java:5]: A team's home ground is also one of it's properties

File: assignment-1/Team.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    public Team(String name, ArrayList<Players> players) {
+        this.Name = name;
+        this.players = players;
+    }
+
+    public Team() {}
+//    public void setName(String name) {
```
Comment (ameyaditya0408) [inline assignment-1/Team.java:13]: Do not leave commented code in your PRs

File: assignment-1/Players.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        return strikeRate;
+    }
+    public int getWickets() {
+        return wickets;
+    }
+
+    public Integer getAllRounders() {
+        return allRounders;
+    }
+}
```
Comment (ameyaditya0408) [inline assignment-1/Players.java:53]: The best practice would be to override the `toString()` method, which would allow us to display the details directly by using the object

File: assignment-1/Players.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    public String getRole() {
+        return role;
+    }
+    public int getMatches() {
+        return matches;
+    }
+    public int getRuns() {
+        return runs;
+    }public Double getAverage() {
```
Comment (ameyaditya0408) [inline assignment-1/Players.java:40]: Format your code properly

File: assignment-1/Players.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
+public class Players {
```
Comment (ameyaditya0408) [inline assignment-1/Players.java:1]: missing hashcode and equals method implementations

File: assignment-1/Team.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+//        return Name;
+//    }
+//
+//    public void setPlayers(ArrayList<Players> players) {
+//        this.players = players;
+//    }
+//    public ArrayList<Players> setPlayers() {
+//        return players;
+//    }
+}
```
Comment (ameyaditya0408) [inline assignment-1/Team.java:27]: If this problem were to be modeled correctly, this class should have methods to return the top players, it's bowlers and all rounders etc, which the main method would call

File: assignment-1/Functions.java [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+    public static ArrayList<Team> allTeams = new ArrayList<Team>();
+    public static <Player> void readData(String csvFile) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:12]: Good work by parameterizing the function

File: assignment-1/Functions.java [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+        try {
+            File file = new File(csvFile);
+            FileReader fr = new FileReader(file);
+            BufferedReader br = new BufferedReader(fr);
+            String line = "";
+            String[] individualField;
+
+            //allPlayers
+
+            while ((line = br.readLine()) != null) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:22]: why would you need 2 loops?
you could have just extracted the header outside the while loop to ignore it

File: assignment-1/Players.java [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    private String name;
+    private String team;
+    private String role;
```
Comment (ameyaditya0408) [inline assignment-1/Players.java:4]: Best practice would be to store the roles as an enum and use the enum constants wherever required instead of plain strings.

File: assignment-1/Functions.java [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+                while ((line = br.readLine()) != null) {
+                    individualField = line.split(delimiter);
+                    String playerName = individualField[0];
+                    String playerTeam = individualField[1];
+                    String playerRole = individualField[2];
+                    int playerMatches = Integer.parseInt(individualField[3]);
+                    int playerRuns = Integer.parseInt(individualField[4]);
+                    Double playerAverage = Double.parseDouble(individualField[5]);
+                    Double playerStrikeRate = Double.parseDouble(individualField[6]);
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:31]: This is susceptible to arrayIndexOutOfBoundException
Should have handled this edge case

File: assignment-1/Functions.java [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+                    int playerWickets = Integer.parseInt(individualField[7]);
+                    allPlayers.add(new Players(playerName, playerTeam, playerRole, playerMatches, playerRuns, playerAverage, playerStrikeRate, playerWickets));
+                }
+            }
+
+            //allTeams object
+
+            for (Players player : allPlayers) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:39]: always ensure that your methods only perform one major operation,
The best practice would have been to move the file operation to a separate utility class that would return a list of csvObjects as response that could be parsed here.

File: assignment-1/Functions.java [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+                boolean flag = false;
+                String temp = player.getTeam();
+                for (Team t : allTeams) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:42]: could have used streams and groupBy to achieve this in a more simple and time-efficient manner.

File: assignment-1/Functions.java [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+                    int playerWickets = Integer.parseInt(individualField[7]);
+                    allPlayers.add(new Players(playerName, playerTeam, playerRole, playerMatches, playerRuns, playerAverage, playerStrikeRate, playerWickets));
+                }
+            }
+
+            //allTeams object
+
+            for (Players player : allPlayers) {
+                boolean flag = false;
+                String temp = player.getTeam();
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:41]: use meaningful variable names

File: assignment-1/Functions.java [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+                for (Team t : allTeams) {
+                    if (t.Name.equals(temp)) {
+                        flag = true;
+                        t.players.add(player);
+                        t.Name = temp;
+                        break;
+                    }
+                }
+                if (!flag) {
+                    ArrayList<Players> play = new ArrayList<Players>();
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:51]: use meaningful variable names

File: assignment-1/Functions.java [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+                }
+            }
+
+            //allTeams object
+
+            for (Players player : allPlayers) {
+                boolean flag = false;
+                String temp = player.getTeam();
+                for (Team t : allTeams) {
+                    if (t.Name.equals(temp)) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:43]: never access variables directly, always use the getter methods.

File: assignment-1/Functions.java [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+                        flag = true;
+                        t.players.add(player);
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:45]: never access variables directly, use setter methods
Create a temporary list of players for the team, and then use the setter methods to add the player. This could have been one possible solution.

File: assignment-1/Functions.java [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+                    }
+                }
+                if (!flag) {
+                    ArrayList<Players> play = new ArrayList<Players>();
+                    play.add(new Players(player.getName(), player.getTeam(), player.getRole(), player.getMatches(), player.getRuns(), player.getAverage(), player.getStrikeRate(), player.getWickets()));
+                    allTeams.add(new Team(temp, play));
+                }
+            }
+
+            br.close();
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:57]: The best practice is to close the resources in the finally block

File: assignment-1/Functions.java [hunk 22]
```diff
@@ -0,0 +1,271 @@
...
+            System.out.println(ioe.getMessage());
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+    }
+
+    //Second Task
+    public static void matchFixture() throws FileNotFoundException {
+        try {
+            PrintWriter pw = new PrintWriter(new File("fixture.csv"));
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:70]: Why read the fixtures from the file again?
you could have modeled the problem better by creating a Fixture class and storing all fixtures as a list. and using that list here.

File: assignment-1/Functions.java [hunk 23]
```diff
@@ -0,0 +1,271 @@
...
+            sb.append("Sr.No");
+            sb.append(",");
+            sb.append("Guest Team");
+            sb.append(",");
+            sb.append(("Home Team"));
+            sb.append(",");
+            sb.append("Venue-Ground");
+            sb.append("\n");
+            Set<String> hash_Set = new HashSet<String>();
+            for (Players t : allPlayers) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:82]: 1. could have retrieved this in an elegant way using filters. or by storing a list of teams while reading the main csv file.
2. or, the better option would have been to use the allteam variable to get the team names.

File: assignment-1/Functions.java [hunk 24]
```diff
@@ -0,0 +1,271 @@
...
+                hash_Set.add(t.getTeam());
+            }
+            Integer srNo = 0;
+            for (String s : hash_Set) {
+                for (String t : hash_Set) {
+                    if (s != t) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:88]: always compare strings using the equals method

File: assignment-1/Functions.java [hunk 25]
```diff
@@ -0,0 +1,271 @@
...
+                        srNo++;
+//                        System.out.println(s + " " + t);
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:90]: do not leave commented code in the PRs

File: assignment-1/Functions.java [hunk 26]
```diff
@@ -0,0 +1,271 @@
...
+                        sb.append(s);
+                        sb.append(",");
+                        sb.append(t);
+                        sb.append(",");
+                        sb.append("null");
+                        sb.append("\n");
+                    }
+                }
+            }
+
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:102]: This whole method should have been segregated into 3 methods performing 3 operations.
1. creation of the fixtures
2. string them in a list
3. writing them to a file

File: assignment-1/Functions.java [hunk 27]
```diff
@@ -0,0 +1,271 @@
...
+                    if (s != t) {
+                        srNo++;
+//                        System.out.println(s + " " + t);
+                        sb.append(srNo);
+                        sb.append(",");
+                        sb.append(s);
+                        sb.append(",");
+                        sb.append(t);
+                        sb.append(",");
+                        sb.append("null");
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:97]: Why null?

File: assignment-1/Functions.java [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+            }
+            Integer srNo = 0;
+            for (String s : hash_Set) {
+                for (String t : hash_Set) {
+                    if (s != t) {
+                        srNo++;
+//                        System.out.println(s + " " + t);
+                        sb.append(srNo);
+                        sb.append(",");
+                        sb.append(s);
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:93]: use better naming convention

File: assignment-1/Functions.java [hunk 29]
```diff
@@ -0,0 +1,271 @@
...
+            System.out.println("fixture.csv file created!");
+
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+
+    }
+
+//    Third Task
+    public static void bowlersWithLeast40Wickets(String bowlersTeam) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:114]: 1. make this method parameterized, so that it is extensible.
2. this method should have been part of the Team class.

File: assignment-1/Functions.java [hunk 30]
```diff
@@ -0,0 +1,271 @@
...
+        try {
+            for (Players p : allPlayers) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:116]: you should have first extracted the team object using the team name, and then used to players list within it to get the top bowlers which would have been a lot more cleaner and understandable.

File: assignment-1/Functions.java [hunk 31]
```diff
@@ -0,0 +1,271 @@
...
+            System.out.println(e.getMessage());
+        }
+    }
+
+    //Fourth Task
+    public static void searchPlayer(String playerName) {
+        try {
+            boolean flag = true;
+            for(Players p : allPlayers) {
+                if(p.getName().contains(playerName))
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:133]: have not handled edge case, where this would fail, since the search is not case insensitive

File: assignment-1/Functions.java [hunk 32]
```diff
@@ -0,0 +1,271 @@
...
+            System.out.println(e.getMessage());
+        }
+
+    }
+
+//    Third Task
+    public static void bowlersWithLeast40Wickets(String bowlersTeam) {
+        try {
+            for (Players p : allPlayers) {
+                if (p.getTeam().equals(bowlersTeam)) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:117]: have not handled edge case, where this would fail, since the search is not case insensitive

File: assignment-1/Functions.java [hunk 33]
```diff
@@ -0,0 +1,271 @@
...
+            for(Players p : allPlayers) {
+                if(p.getName().contains(playerName))
+                {
+                    flag = true;
+                    System.out.println("Name: " + p.getName());
+                    System.out.println("Team: " + p.getTeam());
+                    System.out.println("Role: " + p.getRole());
+                    System.out.println();
+                }
+                else {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:141]: This isn't the correct implementation
consider the situation where the last element of the players list did not contain the keyword, then the flag would be set to false, hence your method would print the list of players and also print that players were not found

File: assignment-1/Functions.java [hunk 34]
```diff
@@ -0,0 +1,271 @@
...
+                System.out.println("No Player found with name " + playerName);
+            }
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+    }
+
+//    Fifth Task
+
+    public static void runsAndWicketsScorer(String teamName) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:155]: This should have been part of the team class

File: assignment-1/Functions.java [hunk 35]
```diff
@@ -0,0 +1,271 @@
...
+
+        HashMap<String, Integer> runScorer = new HashMap<>();
+        HashMap<String, Integer> wicketTaker = new HashMap<>();
+        Map<String, Integer> sorted = new HashMap<>();
+        try {
+            for (Players player : allPlayers) {
+                if (player.getTeam().equals(teamName)) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:162]: do a case-insensitive search

File: assignment-1/Functions.java [hunk 36]
```diff
@@ -0,0 +1,271 @@
...
+                    runScorer.put(player.getName(), player.getRuns());
+                }
+                if (player.getTeam().equals(teamName)) {
+                    wicketTaker.put(player.getName(), player.getWickets());
+                }
+            }
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:168]: Since you already had a list of teams with their players stored, you could directly have used that extracted the top wicket-takers and run scorers from the team. this is a redundant task to generate the structure which already exists.

File: assignment-1/Functions.java [hunk 37]
```diff
@@ -0,0 +1,271 @@
...
+            Map.Entry<String, Integer> entry = sorted.entrySet().iterator().next();
+            System.out.println("Highest Run Scorer");
+            System.out.println(entry.getKey() + " " + entry.getValue());
+
+            sorted = wicketTaker.entrySet().stream().sorted(Collections.reverseOrder(Map.Entry.comparingByValue())).collect(toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e2, LinkedHashMap::new));
+            Map.Entry<String, Integer> entry1 = sorted.entrySet().iterator().next();
+            System.out.println("\nHighest Wicket Taker");
+            System.out.println(entry1.getKey() + " " + entry1.getValue());
+        } catch (Exception e) {
+            e.getMessage();
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:181]: This operation does not do anything!!!

File: assignment-1/Functions.java [hunk 38]
```diff
@@ -0,0 +1,271 @@
...
+            }
+            Map<String, Integer> map = new HashMap<>();
+
+            sorted = runScorer.entrySet().stream().sorted(Collections.reverseOrder(Map.Entry.comparingByValue())).collect(toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e2, LinkedHashMap::new));
+            Map.Entry<String, Integer> entry = sorted.entrySet().iterator().next();
+            System.out.println("Highest Run Scorer");
+            System.out.println(entry.getKey() + " " + entry.getValue());
+
+            sorted = wicketTaker.entrySet().stream().sorted(Collections.reverseOrder(Map.Entry.comparingByValue())).collect(toMap(Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e2, LinkedHashMap::new));
+            Map.Entry<String, Integer> entry1 = sorted.entrySet().iterator().next();
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:177]: use meaningful variable names

File: assignment-1/Functions.java [hunk 39]
```diff
@@ -0,0 +1,271 @@
...
+            System.out.println("\nHighest Wicket Taker");
+            System.out.println(entry1.getKey() + " " + entry1.getValue());
+        } catch (Exception e) {
+            e.getMessage();
+        }
+    }
+
+    //Sixth Task
+    public static void top3BatBowlAllRounder() {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:186]: make this method parameterized, so that it is extensible.
This should have been part of the Team class

File: assignment-1/Functions.java [hunk 40]
```diff
@@ -0,0 +1,271 @@
...
+        System.out.println("\nTop 3 Batsman of the season");
+        try {
+            List<Players> topBatters = allPlayers.stream()
+                    .sorted(Comparator.comparingInt(Players::getRuns).reversed()).limit(3)
+                    .collect(Collectors.toList());
+            for (Players lp : topBatters) {
+                System.out.println(lp.getName());
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:193]: if you had overridden the toString method, you could have directly printed the player object here

File: assignment-1/Players.java [hunk 41]
```diff
@@ -0,0 +1,53 @@
...
+    public Players(String name, String team, String role, Integer matches, Integer runs, Double average, Double strikeRate, Integer wickets) {
+        this.name = name;
+        this.team = team;
+        this.role = role;
+        this.matches = matches;
+        this.runs = runs;
+        this.average = average;
+        this.strikeRate = strikeRate;
+        this.wickets = wickets;
+        this.allRounders = wickets*runs;
```
Comment (ameyaditya0408) [inline assignment-1/Players.java:22]: this is only applicable if the role is an All-rounder. but since there is no distinction, this will be applied to both bowlers and batsmen.
it should have made sense to model your player class such that you have an AllrounderPlayer which extends from the Player class and implement this method only this the sub class.

File: assignment-1/Functions.java [hunk 42]
```diff
@@ -0,0 +1,271 @@
...
+        }
+    }
+
+    public static void highestRunScorerTeam() {
+        HashMap<String, Double> teamAndAvg = new HashMap<>();
+        LinkedHashMap<String, Double> reverseSorted = new LinkedHashMap<>();
+        try {
+            for (Team t : allTeams) {
+                List<Players> topBatsman = t.players.stream()
+                        .sorted(Comparator.comparingDouble(Players::getAverage).reversed()).limit(6)
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:221]: why 6? didn't the requirement say, you should have 11 players?

File: assignment-1/Functions.java [hunk 43]
```diff
@@ -0,0 +1,271 @@
...
+                        .collect(Collectors.toList());
+                Double score = topBatsman.stream().mapToDouble(Players::getAverage).sum();
+                teamAndAvg.put(t.Name, score);
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:224]: always use getter methods

File: assignment-1/Functions.java [hunk 44]
```diff
@@ -0,0 +1,271 @@
...
+            for (Players lp : topAllrounders) {
+                System.out.println(lp.getName());
+            }
+
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        }
+    }
+
+    public static void highestRunScorerTeam() {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:215]: Since this was an opened question, you were expected to add an explanation on the criteria you followed to choose the highest scorers and top teams

File: assignment-1/Functions.java [hunk 45]
```diff
@@ -0,0 +1,271 @@
...
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+        } finally {
+            System.out.println("________________________________________________________");
+        }
+    }
+
+
+    // Eighth Task
+    public static void nextGenPlayers() {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:248]: Explain the qualifying criteria for a nexgenPlayer

File: assignment-1/Functions.java [hunk 46]
```diff
@@ -0,0 +1,271 @@
...
+                        nextGenPlayers.add(newGenPlayer);
+                    }
+                }
+                if (newGenPlayer.getMatches() <= 15) {
+                    if (newGenPlayer.getWickets() >= 10) {
+                        nextGenPlayers.add(newGenPlayer);
+                    }
+                }
+            }
+            for (Players newPlayers : nextGenPlayers) {
```
Comment (ameyaditya0408) [inline assignment-1/Functions.java:264]: This feels like an incomplete task, since the question was asked to find nexgen players for all teams,
I do not see any team names being printed

**Review by ameyaditya0408 — COMMENTED** [review]
1. OOP modeling was not up to the mark, expected you define more meaning full classes and establish relationships among them.
2. Wanted to see the use of custom exceptions being thrown wherever necessary.
3. Structure the project into meaningful files and packages, instead of adding everything in one file
4. Functions could have been parameterized to make them more extensible.
5. Edge cases and null pointer exceptions are not handled in a few functions.
6. Always add documentation to public classes and methods.
7. Did not see any test cases being written.
8.
